File: back-end/src/main/java/es/ull/project/domain/enumerate/TimeUnit.java
package es.ull.project.domain.enumerate;


import java.util.Random;

/**
 * TimeUnit
 *
 * Represents the different time units used for temporal measurements.
 * Each unit defines a specific time period granularity.
 * DAY, WEEK, MONTH, or YEAR.
 */
public enum TimeUnit {

    /**
     * Day time unit.
     */
    DAY,

    /**
     * Week time unit.
     */
    WEEK,

    /**
     * Month time unit.
     */
    MONTH,

    /**
     * Year time unit.
     */
    YEAR;

    /**
     * Returns the TimeUnit that matches the given string.
     *
     * @param stringToCheck String representation of the time unit.
     * @return TimeUnit matching the string.
     * @throws IllegalArgumentException if the string is null or invalid.
     */
    public static TimeUnit fromString(String stringToCheck) {
        if (stringToCheck == null) {
            throw new IllegalArgumentException("Time unit is not defined");
        }

        stringToCheck = stringToCheck.trim().toUpperCase();

        for (TimeUnit timeUnit : values()) {
            if (timeUnit.name().equals(stringToCheck)) {
                return timeUnit;
            }
        }

        throw new IllegalArgumentException("Invalid time unit");
    }

    /**
     * Checks whether the given string is a valid TimeUnit.
     *
     * @param stringToCheck String to validate.
     * @return True if valid, false otherwise.
     */
    public static boolean isValid(String stringToCheck) {
        if (stringToCheck == null) {
            return false;
        }

        stringToCheck = stringToCheck.trim().toUpperCase();

        for (TimeUnit timeUnit : values()) {
            if (timeUnit.name().equals(stringToCheck)) {
                return true;
            }
        }
        return false;
    }

    /**
     * Returns a random TimeUnit.
     * Useful for testing purposes.
     *
     * @return Random TimeUnit.
     */
    public static TimeUnit random() {
        return values()[new Random().nextInt(values().length)];
    }


}

File: front-end/src/domain/enumerate/time-unit.ts
/**
 * TimeUnit
 *
 * Represents the different time units used for temporal measurements.
 *
 * Values:
 * - DAY: Day time unit.
 * - WEEK: Week time unit.
 * - MONTH: Month time unit.
 * - YEAR: Year time unit.
 */
export enum TimeUnit {
  /** Day time unit. */
  DAY = 'DAY',
  /** Week time unit. */
  WEEK = 'WEEK',
  /** Month time unit. */
  MONTH = 'MONTH',
  /** Year time unit. */
  YEAR = 'YEAR',
}

/**
 * Returns an array with all TimeUnit values.
 * @returns {TimeUnit[]} array with allowed time units.
 */
export function timeUnitValues(): TimeUnit[] {
  return Object.values(TimeUnit) as TimeUnit[];
}

/**
 * Parse a string into TimeUnit.
 * @param {string} [s] String representation of the time unit.
 * @returns {TimeUnit} matching time unit.
 * @throws {Error} if undefined or invalid.
 */
export function timeUnitFromString(s?: string): TimeUnit {
  if (!s) throw new Error('Time unit is not defined');
  const key = s.trim().toUpperCase();
  for (const v of timeUnitValues()) {
    if (v === key) return v;
  }
  throw new Error('Invalid time unit');
}

/**
 * Checks whether the given string is a valid TimeUnit.
 * @param {string} [s] String to validate.
 * @returns {boolean} True if valid, false otherwise.
 */
export function isTimeUnit(s?: string): boolean {
  if (!s) return false;
  const key = s.trim().toUpperCase();
  for (const v of timeUnitValues()) {
    if (v === key) return true;
  }
  return false;
}

/**
 * Return a random TimeUnit (useful for testing).
 * @returns {TimeUnit} random time unit.
 */
export function timeUnitRandom(): TimeUnit {
  const values = timeUnitValues();
  return values[Math.floor(Math.random() * values.length)]!;
}

/**
 * Returns the options for a time unit selector.
 * @param {Function} t Translation function.
 * @returns {{ title: string; value: TimeUnit }[]} array of options.
 */
export function timeUnitToOptions(t: (key: string) => string): { title: string; value: TimeUnit }[] {
  return timeUnitValues().map((unit) => ({
    title: t(`common.timeUnits.${unit}`),
    value: unit,
  }));
}
